src/hotspot/share/utilities/vmError.cpp

```diff
@@ -1312,6 +1313,38 @@ int VMError::prepare_log_file(const char* pattern, const char* default_pattern,
   return fd;
 }
 
+// Attempt to set the state to Native in VMError.
+// By setting it to Native, it allows os::fork_and_exec to execute cmd such as jcmd %p.
+// Otherwise, it may end up with a deadlock when cmd tries to synchronize all Java threads
+// at safepoints.
+class VMErrorThreadToNativeFromVM : public StackObj {
+ private:
+   JavaThread* _thread;
+
+ public:
+  VMErrorThreadToNativeFromVM(Thread* t) : _thread(nullptr) {
+    if (t != nullptr && t->is_Java_thread() &&
+        !Threads_lock->owned_by_self()) { // VMError::controlled_crash() grabs Threads_lock.
+      _thread = JavaThread::cast(t);
+      assert(_thread == Thread::current(), "must be current thread");
+      assert(_thread->thread_state() == _thread_in_vm, "must be in VM");
```

Comment:
Don't assert here. We are in the middle of error handling. This would just lead to recursive errors and very probably to "too many errors, abort".

Reply:
Acknowledge! I will avoid asserts and reverting the state back.